merlin/models/torch/models/base.py

```diff
@@ -0,0 +1,226 @@
+#
+from typing import Dict, List, Optional, Sequence, Union
+
+import torch
+from pytorch_lightning import LightningModule
+from torch import nn
+
+from merlin.dataloader.torch import Loader
+from merlin.io import Dataset
+from merlin.models.torch.batch import Batch, sample_batch
+from merlin.models.torch.block import Block
+from merlin.models.torch.container import BlockContainer
+from merlin.models.torch.outputs.base import ModelOutput
+from merlin.models.torch.utils import module_utils
+from merlin.models.utils.registry import camelcase_to_snakecase
+from merlin.schema import Schema
+
+
+class Model(Block, LightningModule):
+    """
+    Merlin Model class.
+
+    The Model class extends from both the Block and LightningModule classes. It
+    allows for easy construction of models using pre-defined blocks.
+
+    Parameters
+    ----------
+    *blocks: nn.Module
+        One or more blocks that make up the core functionality of the model.
+    schema: Schema, optional
+        A Merlin schema. Default is None.
+    optimizer: torch.optim.Optimizer, optional
+        A PyTorch optimizer from the PyTorch library (or any custom optimizer
+        that follows the same API). Default is Adam optimizer.
+
+    Example usage
+    -------------
+    >>> model = Model(
+    ...    TabularInputBlock(schema),
+    ...    MLPBlock([32, 16]),
+    ...    BinaryOutput(schema.select_by_tag(Tags.TARGET).first),
+    ... )
+    ... trainer = Trainer(max_epochs=1)
+    ... with Loader(dataset, batch_size=16) as loader:
+    ...     model.initialize(loader)
+    ...     trainer.fit(model, loader)
+    """
+
+    def __init__(
+        self,
+        *blocks: nn.Module,
+        schema: Optional[Schema] = None,
+        optimizer=torch.optim.Adam,
+    ):
+        """Initializes `Model` class"""
+        super().__init__()
+        self.schema = schema
+
+        self.pre = BlockContainer(name="pre")
+        self.blocks = BlockContainer(name="blocks")
+        for block in blocks:
+            self.blocks.append(block)
+        self.post = BlockContainer(name="post")
+
+        self.optimizer = optimizer
+
+    def initialize(self, data: Union[Dataset, Loader, Batch]):
+        """Initializes the model based on a given data set."""
+        return initialize(self, data)
+
+    def forward(
+        self, inputs: Union[torch.Tensor, Dict[str, torch.Tensor]], batch: Optional[Batch] = None
+    ):
+        """Performs a forward pass through the model."""
+        outputs = inputs
+        for pre in self.pre.values:
+            outputs = pre(outputs, batch=batch)
+        for block in self.blocks.values:
+            outputs = block(outputs, batch=batch)
+        for post in self.post.values:
+            outputs = post(outputs, batch=batch)
+        return outputs
+
+    def training_step(self, batch, batch_idx):
+        """Performs a training step with a single batch."""
+        del batch_idx
+        inputs, targets = batch
```

Comment:
We need some logic here to construct the `Batch` class and pass it to the forward-pass.